### DISCUSSION TITLE
approach to use a context tag

### DISCUSSION THREAD

**Discussion by jwinarske** [gabime/spdlog#3135] [General]
I'm implementing a lower layer API (DLT) that wants a unique context tag independent of the message.  The DLT API requires two logical calls to set:
1. context tag (only on state change)
2. message

Upper level user would be able to set context tag ad-hoc, independent of message.  I'm currently using `callback_logger_mt` to shim the DLT message API.

Aside from encoding message with a unique token for the `context tag`, is there a suggested way to do this with the spdlog api?

My current implementation:
https://github.com/toyota-connected/ivi-homescreen/tree/v2.0/shell/logging

Reference:
https://covesa.github.io/dlt-daemon

**Comment by jwinarske** [comment]
This scenario (custom attributes) looks interesting:  https://github.com/gabime/spdlog/issues/673.  Although I would make them optional with a default of nullptr.

**Comment by jwinarske** [comment]
Also custom attributes would only make sense with a custom sink

**Comment by tt4g** [comment]
There are PRs with similar features #2907, #3128, but none that refer to values in the sink.
You should use custom sinks as they are implemented now.

**Reply by jwinarske** [reply]
#2907 does not support async
#3128 is not merged

I aggregate logs from multiple threads to a custom ring buffer sink, then pass to global logger calling DLT using callback_logger_mt.

The context can change for each message.  How would you propose to associate the MDC key/value and message to reference in custom sink?

**Reply by tt4g** [reply]
If it changes for each message, I would think that it would be better to simply pass the context in the logging argument.

**Reply by jwinarske** [reply]
This is what I like about https://github.com/gabime/spdlog/issues/673

**Reply by tt4g** [reply]
You cannot pass arguments to sink via loggers.
You must manage a custom sink paired with a logger and devise a way to pass arguments to the sink.

Alternatively, you can create a user-defined formatter to format custom arguments.
See: https://fmt.dev/10.2.0/api.html#formatting-user-defined-types
